legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/main/java/org/finos/legend/engine/persistence/components/relational/h2/H2Sink.java

```diff
@@ -236,4 +261,86 @@ public IngestorResult performBulkLoad(Datasets datasets, Executor<SqlGen, Tabula
 
         return result;
     }
+
+    public List<DataError> performDryRun(Datasets datasets, Transformer<SqlGen, SqlPlan> transformer, Executor<SqlGen, TabularData, SqlPlan> executor, SqlPlan dryRunSqlPlan, Map<ValidationCategory, Map<Set<FieldValue>, SqlPlan>> dryRunValidationSqlPlan, int sampleRowCount)
+    {
+        executor.executePhysicalPlan(dryRunSqlPlan);
+
+        List<DataError> dataErrors = new ArrayList<>();
+        Map<ValidationCategory, Queue<DataError>> dataErrorsByCategory = new HashMap<>();
+        for (ValidationCategory validationCategory : ValidationCategory.values())
+        {
+            dataErrorsByCategory.put(validationCategory, new LinkedList<>());
+        }
+
+        List<String> allFields = datasets.stagingDataset().schemaReference().fieldValues().stream().map(FieldValue::fieldName).collect(Collectors.toList());
+
+        Map<Set<FieldValue>, SqlPlan> queriesForNull =  dryRunValidationSqlPlan.getOrDefault(NULL_VALUES, new HashMap<>());
+        Map<Set<FieldValue>, SqlPlan> queriesForDatatype = dryRunValidationSqlPlan.getOrDefault(DATATYPE_CONVERSION, new HashMap<>());
+
+        // Execute queries for null values
+        for (Set<FieldValue> validatedColumns : queriesForNull.keySet())
+        {
+            List<TabularData> results = executor.executePhysicalPlanAndGetResults(queriesForNull.get(validatedColumns));
+            if (!results.isEmpty())
+            {
+                List<Map<String, Object>> resultSets = results.get(0).getData();
+                for (Map<String, Object> row : resultSets)
+                {
+                    for (String column : validatedColumns.stream().map(FieldValue::fieldName).collect(Collectors.toSet()))
+                    {
+                        if (row.get(column) == null)
+                        {
+                            DataError dataError = constructDataError(allFields, row, FILE, ROW_NUMBER, NULL_VALUES, column);
+                            dataErrors.add(dataError);
+                            dataErrorsByCategory.get(NULL_VALUES).add(dataError);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Execute queries for datatype conversion
+        for (Set<FieldValue> validatedColumns : queriesForDatatype.keySet())
+        {
+            try
+            {
+                executor.executePhysicalPlanAndGetResults(queriesForDatatype.get(validatedColumns));
+            }
+            catch (RuntimeException e)
+            {
+                String errorMessage = e.getCause().getMessage();
+                String problematicValue = errorMessage.substring(0, errorMessage.indexOf("; SQL statement"));
+                problematicValue = problematicValue.replace("org.h2.jdbc.JdbcSQLDataException: Data conversion error converting ", "");
+                problematicValue = problematicValue.replace("org.h2.jdbc.JdbcSQLDataException: Cannot parse \"TIMESTAMP\" constant ", "");
```

Comment:
Let's explore if using regex to extract value will be more helpful

Reply:
```
        if (errorMessage.contains("Data conversion error"))
        { ...        }
        else if (errorMessage.contains("Cannot parse"))
        {
           ....
        }
        return errorMessage;
```

what shall we do if error Message does not fall in "if block" or "else if block"  ?
I see we are just returning it atm.